Alerting: Fix notification policies inheritance algorithm

gillesdemey opened the pull request:
**What is this feature?**

This PR fixes the inheritance algorithm to compute what properties are inherited from a parent policy in a child policy.

**Additionally** 

This PR normalizes the behaviour of the notification policies API response between Mimir and the built-in Alertmanager / vanilla Alertmanager. 

It also changes to way we write the config for `inherit from parent`, `group_by: []` is only preserved in the configuration by Mimir – other implementations simply return `undefined`.

---

## Note on grouping

There are several ways to define the `group_by` property of a notification policy, the following describes the values and their behaviour.

* `group_by: ['label']` is just regular grouping
* `group_by: []` means "inherit from parent" for child policies
* `group_by: []` means "single group" for the default (root) policy
* `group_by: <undefined>` (the absence of the `group_by` key in the policy) inherits the grouping from the parent route
* `group_by: ['...']` means disable grouping

**Special notes for your reviewer:**

Please review the changes in the inheritance algorithm more closely than the rest of the PR – which is mostly updating the UI to take "no grouping" and "single group" semantics in to account.

I also moved that code [to this file](https://github.com/grafana/grafana/pull/69304/files#diff-226473adaf086beab36f8ac2cd83047a80b10e95657ed68d777debd686df3cf0
)

gillesdemey committed:
wip fix policies empty array

gillesdemey committed:
add support for null values in policy inheritance

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -107,10 +107,15 @@ const Policy: FC<PolicyComponentProps> = ({
+
+  const singleGroup = !noGrouping && isArray(groupBy) && groupBy.length === 0;
+  const customGrouping = !noGrouping && isArray(groupBy) && groupBy.length > 0;
+
```
☝️ nice!

soniaAguilarPeiron reviewed (approved):
LGTM!🚀

soniaAguilarPeiron commented on public/app/features/alerting/unified/utils/amroutes.ts:
```diff
@@ -183,7 +183,7 @@ export const formAmRouteToAmRoute = (
   } = formAmRoute;
 
-  const group_by = overrideGrouping && groupBy ? groupBy : [];
+  const group_by = overrideGrouping ? (groupBy ? groupBy : []) : undefined;
```
i would add a comment in here, to make more clear we are setting undefined to inherit from parent.
```suggestion
  const inheritFromParent= undefined;
  const group_by = overrideGrouping ? (groupBy ? groupBy : []) : inheritFromParent;
```

gillesdemey committed:
update tests

gillesdemey committed:
PR feedback

gillesdemey commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -218,17 +223,27 @@ const Policy: FC<PolicyComponentProps> = ({
+                      <Strong>Single group</Strong>
+                    </MetaText>
+                  )}
+                  {/* we only want to show "no grouping" on the root policy, children with empty groupBy will inherit from the parent policy */}
```
Yep, it's no longer accurate – removed

gillesdemey committed:
fix BE test

gillesdemey committed:
disallow empty group_by for now

since the behaviour is currently inconsistent between AM flavors and we're not sure which one is correct

gillesdemey commented on public/app/features/alerting/unified/components/Strong.tsx:
```diff
@@ -6,7 +6,7 @@ interface Props {}
 const Strong = ({ children }: React.PropsWithChildren<Props>) => {
   const theme = useTheme2();
-  return <strong style={{ color: theme.colors.text.maxContrast }}>{children}</strong>;
+  return <strong style={{ color: theme.colors.text.primary }}>{children}</strong>;
```
unrelated to the fix, but looks better :)

gillesdemey committed:
some small updates

gillesdemey committed:
support both empty array and undefined as group_by inherit value

gillesdemey committed:
fix: use correct inheritance for group_by: []

gillesdemey committed:
add tests, update algorithm for bug

gillesdemey committed:
Merge branch 'main' into alerting/fix-notification-policies-grouping-copy

gillesdemey committed:
revert drone file

not sure how this one got updated

gillesdemey committed:
simplify code

soniaAguilarPeiron commented on public/app/features/alerting/unified/utils/amroutes.ts:
```diff
@@ -108,8 +108,8 @@ export const amRouteToFormAmRoute = (route: RouteWithID | Route | undefined): Fo
-    overrideGrouping: Array.isArray(route.group_by) && route.group_by.length !== 0,
-    groupBy: route.group_by ?? [],
+    overrideGrouping: Array.isArray(route.group_by) && route.group_by.length > 0,
+    groupBy: route.group_by ?? undefined,
```
Could we use the new constant `INHERIT_FROM_PARENT` in here instead of undefined?

soniaAguilarPeiron reviewed (approved):
LGTM!🚀

gillesdemey commented:
Back-port PR here 👉 https://github.com/grafana/grafana/pull/69782
